util/overlay/overlay_linux.go

```diff
@@ -322,3 +351,55 @@ func compareCapabilities(p1, p2 string) (bool, error) {
 	}
 	return bytes.Equal(c1, c2), nil
 }
+
+// Ported from continuity project
+// https://github.com/containerd/continuity/blob/bce1c3f9669b6f3e7f6656ee715b0b4d75fa64a6/fs/path.go#L135
+// Copyright The containerd Authors.
+func compareSymlinkTarget(p1, p2 string) (bool, error) {
+	t1, err := os.Readlink(p1)
+	if err != nil {
+		return false, err
+	}
+	t2, err := os.Readlink(p2)
+	if err != nil {
+		return false, err
+	}
+	return t1 == t2, nil
+}
+
+const compareChunkSize = 32 * 1024
+
+// Ported from continuity project
+// https://github.com/containerd/continuity/blob/bce1c3f9669b6f3e7f6656ee715b0b4d75fa64a6/fs/path.go#L151
+// Copyright The containerd Authors.
+func compareFileContent(p1, p2 string) (bool, error) {
+	f1, err := os.Open(p1)
+	if err != nil {
+		return false, err
+	}
+	defer f1.Close()
+	f2, err := os.Open(p2)
+	if err != nil {
+		return false, err
+	}
+	defer f2.Close()
+
+	b1 := make([]byte, compareChunkSize)
+	b2 := make([]byte, compareChunkSize)
+	for {
+		n1, err1 := f1.Read(b1)
+		if err1 != nil && err1 != io.EOF {
+			return false, err1
+		}
+		n2, err2 := f2.Read(b2)
+		if err2 != nil && err2 != io.EOF {
+			return false, err2
+		}
+		if n1 != n2 || !bytes.Equal(b1[:n1], b2[:n2]) {
```

Comment:
how is the `n1 != n2` comparison correct in here?

Reply:
There's no guarantee that `Read` returns a specific number of bytes. It just returns whatever single syscall returned.

Reply:
Yes, and the `read` syscall is made with `count` parameter set to the `len` of the provided buffer, which is the same for both files, so the situations where `n1 != n2` are either:
1. The files aren't the same size (in which case they aren't equal)
2. They aren't regular files (which we guard against)
3. The syscall was interrupted (which should be handled by the go runtime or otherwise result in a non-nil `err` being returned)
4. The underlying filesystem is behaving extremely strangely and returning less than `count` bytes for a regular file even though `EOF` hasn't been reached and nothing interrupted it.

We can probably turn `4` into an error condition rather than just returning `false, nil`, but I'd prefer to do that along with updating `continuity` to have the same behavior so the differs are consistent.

Reply:
> behaving extremely strangely 

I don't think this case is extremely strange. It's part of the contract. Network-based filesystems definitely do it. I have not looked at the internals of every possible linux filesystem and even when some of them guarantee it atm they can change it any time. There is no guarantee for it even in the vfs-to-syscall layer iiuc.

Reply:
Can't we use io.ReadFull here?

Reply:
Yeah, I always thought the rule of thumb was that regular files only return short reads on EOF, signal interrupt or conditions where `errno` is set, but you're right that it's not technically in violation of the spec to do otherwise and network filesystems or FUSE filesystems are much more likely to do things outside the norm, so we should handle this under the assumption that eventually someone will do something very weird (like you've said previously).

> Can't we use io.ReadFull here?
Yeah good call, that handles the complication for us. Updated with that. I will send out a PR to continuity to have the double-walking differ do the same thing since that's where this code was originally from.

Reply:
By the way, I left in `n1 != n2` in the update because it should be correct now that `ReadFull` is used. It's technically also asserted by `byte.Equal` but I'm not sure how efficiently so I figured it didn't hurt to leave it here.